File: mpv/scripts/ankiminer/anki.lua
local util = require('mp.utils')
local media = require("media")
local time_to_string = require("utils").time_to_string

local M = {}

-- All anki requests are of the form:
-- {
--  "action" : @action
--  "version" : 6
--  "params" : @t_params (optional)
-- }
-- and return a lua table with 2 keys, "result" and "error"
function M.anki_connect(action, t_params, url)
  local url = url or 'localhost:8765'

  local request = {
    action = action,
    version = 6,
  }

  if t_params ~= nil then
    request.params = t_params
  end

  local json = util.format_json(request)

  local command = {
    'curl', url, '-X', 'POST', '-d', json
  }

  local result = mp.command_native({
    name = 'subprocess',
    args = command,
    capture_stdout = true,
    capture_stderr = true,
  })

  result = util.parse_json(result.stdout)

  return result.result, result.error
end


function M.create_anki_note(gui, options, ctx)
  local filename_prefix = media.generate_media_name()

  local filename_audio = media.create_audio(
    options.media_path,
    filename_prefix,
    ctx.start_time,
    ctx.end_time,
    options.audio_bitrate)

  local filename_snapshot = media.create_snapshot(
    options.media_path,
    filename_prefix,
    ctx.snapshot_time,
    options.snapshot_height
  )
  print("filename_snapshot: ", filename_snapshot)

  -- Start filling the fields
  local fields = {}

  if #filename_audio > 0 then
    fields[options.field_audio] = '[sound:' .. filename_audio .. ']'
  end

  if #filename_snapshot > 0 then
    fields[options.field_snapshot] = '<img src="' .. filename_snapshot .. '">'
  end

  fields[options.field_subtitle1] = table.concat(ctx.subtitles.main, "\n")
  fields[options.field_subtitle2] = table.concat(ctx.subtitles.secondary, "\n")
  fields[options.field_title] = ctx.word

  local param = {
    note = {
      deckName = options.deckname,
      modelName = options.modelName,
      fields = fields,
      tags = {
        'mpv2anki'
      }
    }
  }

  local action;

  if gui then
    action = 'guiAddCards'
  else
    action = 'addNote'
  end

  inspect(param)

  M.anki_connect(action, param, options.anki_url)
end


return M

File: mpv/scripts/ankits/src/lib/anki.ts
import { Config, RecordingContext } from "../types";
import { createAudio, createSnapshot, generate_media_name } from "./media";

export function anki_connect(opts: {
  action: string,
  t_params: any,
  url: string,
}): void {

  let req = {
    action: opts.action,
    version: 6,
    params: opts.t_params,
  }

  let json = JSON.stringify(req);
  let cmd = ["curl", opts.url, "-X", "POST", "-d", json]
  let result = mp.command_native({
    name: 'subprocess',
    args: cmd,
    capture_stdout: true,
    capture_stderr: true,
  })

  let response = JSON.parse(result.stdout);

  return response.result, response.error

}

export function create_anki_note(cfg: Config, ctx: RecordingContext): void {
  let filename_prefix = generate_media_name();

  let filename_audio = createAudio({
    mediaPath: cfg.media_path,
    filenamePrefix: filename_prefix,
    startTime: ctx.start_time,
    endTime: ctx.end_time,
  })

  let filename_snapshot = createSnapshot({
    mediaPath: cfg.media_path,
    filenamePrefix: filename_prefix,
    snapshotTime: ctx.snapshot_time,
  })
  print(`filename snapshot: ${filename_snapshot}`);
  let fields = {};
  fields[cfg.field_audio] = `[sound:${filename_audio}]`;
  fields[cfg.field_snapshot] = `<img src="${filename_snapshot}">`;
  fields[cfg.field_subtitle1] = ctx.subtitles.main.join("\n");
  fields[cfg.field_subtitle2] = ctx.subtitles.secondary.join("\n");
  fields[cfg.field_title] = "word not defiend yet";

  let params = {
    note: {
      deckName: cfg.deckname,
      modelName: cfg.modelName,
      fields: fields,
    }
  }

  print("ready to export http post: ");
  dump(params);

  anki_connect({
    action: "guiAddCards",
    t_params: params,
    url: cfg.anki_url,
  })

}





